### PR TITLE
Multiplication overflow detection for types of size more than 64 bits.

### PR DESCRIPTION
**Pull request by salaheldinsoliman**
Multiplication overflow detection for bits > 64.

### CODE REVIEW

File: tests/substrate_tests/expressions.rs [hunk 1]
```diff
@@ -1000,7 +1004,6 @@ fn multiply() {
         assert_eq!(res, runtime.vm.output);
     }
 }
-
 #[test]
```
Comment (seanyoung) [inline tests/substrate_tests/expressions.rs:1007]: please don't remove random lines :)

File: tests/solana_tests/primitives.rs [hunk 2]
```diff
@@ -662,6 +878,9 @@ fn int() {
             rng.fill(&mut a_bs[..]);
             rng.fill(&mut b_bs[..]);
 
+            println!("a : {:?}", a_bs);
+            println!("b : {:?}", b_bs);
+
             let mut a = U256::from_big_endian(&a_bs);
```
Comment (seanyoung) [inline tests/solana_tests/primitives.rs:884]: Please remove debugging statements

File: tests/solana_tests/primitives.rs [hunk 3]
```diff
@@ -599,6 +604,215 @@ fn truncate_uint(n: &mut U256, width: usize) {
...
+         
+            function mul(uintN a, uintN b) public returns (uintN) {
+                
+                return a * b;
+                
+            }
+
+           
+
+        }"#
```
Comment (seanyoung) [inline tests/solana_tests/primitives.rs:786]: lots of unnecessary blank lines

File: tests/solana_tests/primitives.rs [hunk 4]
```diff
@@ -599,6 +604,215 @@ fn truncate_uint(n: &mut U256, width: usize) {
...
+
+        assert_ne!(res, Ok(0));
+    }
+}
+
+#[test]
+
+fn test_overflow_detect_unsigned() {
+    let mut rng = rand::thread_rng();
+    for width in (72..=256).step_by(8) {
```
Comment (seanyoung) [inline tests/solana_tests/primitives.rs:774]: what about 8..64?

File: tests/solana_tests/primitives.rs [hunk 5]
```diff
@@ -599,6 +604,215 @@ fn truncate_uint(n: &mut U256, width: usize) {
...
+            let res = contract.function_must_fail(
+                "mul",
+                &[
+                    ethabi::Token::Uint(U256::from_big_endian(&first_operand_rand.to_bytes_be())),
+                    ethabi::Token::Uint(U256::from_big_endian(&second_operand_rand.to_bytes_be())),
+                ],
+                &[],
+                None,
+            );
+            assert_ne!(res, Ok(0));
```
Comment (seanyoung) [inline tests/solana_tests/primitives.rs:811]: also add a test for a value which does not overflow

File: tests/solana_tests/primitives.rs [hunk 6]
```diff
@@ -599,6 +604,215 @@ fn truncate_uint(n: &mut U256, width: usize) {
...
+         
+            function mul(intN a, intN b) public returns (intN) {
+                
+                return a * b;
+                
+            }
+
+           
+
+        }"#
```
Comment (seanyoung) [inline tests/solana_tests/primitives.rs:727]: remove blank lines

File: src/emit/mod.rs [hunk 7]
```diff
@@ -5505,6 +5505,200 @@ pub trait TargetRuntime<'a> {
...
+        l: PointerValue<'a>,
+        r: PointerValue<'a>,
+        o: PointerValue<'a>,
+        mul_bits: u32,
+        left: IntValue<'a>,
+        right: IntValue<'a>,
+        bits: u32,
+        function: FunctionValue,
+    ) -> IntValue<'a> {
+        // In llvm, Two's complement representation is used.
```
Comment (seanyoung) [inline src/emit/mod.rs:5520]: Not just llvm. Solidity uses two's complement and I think all modern cpu isas

File: src/emit/mod.rs [hunk 8]
```diff
@@ -5505,6 +5505,200 @@ pub trait TargetRuntime<'a> {
...
+        // We check for overflow based on the facts:
+        //  - * - = +
+        //  + * + = +
+        //  - * + = - (if op1 or op2 != 0)
```
Comment (seanyoung) [inline src/emit/mod.rs:5524]: do you mean `op1 != 0 and op2 != 0`?

File: src/emit/mod.rs [hunk 9]
```diff
@@ -5505,6 +5505,200 @@ pub trait TargetRuntime<'a> {
...
+            self.extract_sign_bit(
+                bin,
+                bin.builder
+                    .build_int_truncate(res.into_int_value(), left.get_type(), ""),
+                left.get_type(),
+            )
+        };
+
+        // If the operands sign is the same
+        let similar_operands =
```
Comment (seanyoung) [inline src/emit/mod.rs:5580]: `similar_operands` is not great. How about `same_signs`?

File: src/emit/mod.rs [hunk 10]
```diff
@@ -5505,6 +5505,200 @@ pub trait TargetRuntime<'a> {
...
+            bin.builder
+                .build_int_compare(IntPredicate::EQ, left_sign_bit, right_sign_bit, "");
+
+        // If operands sign is the same and the result is positive
+        let sim_operands_and_positive_res = bin.builder.build_and(
```
Comment (seanyoung) [inline src/emit/mod.rs:5585]: `same_` ?

File: src/emit/mod.rs [hunk 11]
```diff
@@ -5505,6 +5505,200 @@ pub trait TargetRuntime<'a> {
...
+    ) -> IntValue<'a> {
+        let n_bits_to_shift = int_type.get_bit_width() - 1;
+        let val_to_shift = int_type.const_int(n_bits_to_shift as u64, false);
+        let shifted = bin
+            .builder
+            .build_right_shift(operand, val_to_shift, false, "");
+        bin.builder
+            .build_int_truncate(shifted, bin.context.bool_type(), "")
+    }
     // emit a multiply for any width with or without overflow checking
```
Comment (seanyoung) [inline src/emit/mod.rs:5702]: empty line after function  body please

File: src/emit/mod.rs [hunk 12]
```diff
@@ -5722,7 +5982,6 @@ pub trait TargetRuntime<'a> {
 
         function
     }
-
     /// Convenience function for generating binary operations with overflow checking.
```
Comment (seanyoung) [inline src/emit/mod.rs:5985]: please don't remove random lines

**Comment by LucasSte** [comment]
Please, prove a explanatory name for your PR. `Dev` does not sound good.

**Review by seanyoung — CHANGES_REQUESTED** [review]
Looks good but there details to deal with, see the comments.

The pull request needs a better name.

Also have you written tests for compile-time overflow tests with constants?

File: Cargo.toml [hunk 13]
```diff
@@ -55,7 +56,7 @@ num-derive = "0.3"
 parity-scale-codec = "3.1"
 ethabi = "17.0"
 wasmi = "0.11"
-rand = "0.7"
+rand_07 = { package = "rand", version = "0.7" }
```
Comment (LucasSte) [inline Cargo.toml:59]: Why did you change this? You already have `rand = 0.8` here.

File: src/emit/mod.rs [hunk 14]
```diff
@@ -5505,6 +5505,200 @@ pub trait TargetRuntime<'a> {
...
+        let res = bin.builder.build_load(o, "mul");
+
+        let error_block = bin.context.append_basic_block(function, "error");
+        let return_block = bin.context.append_basic_block(function, "return_block");
+
+        // Extract sign bits of the operands and the result
+        let left_sign_bit = self.extract_sign_bit(bin, left, left.get_type());
+        let right_sign_bit = self.extract_sign_bit(bin, right, right.get_type());
+        let res_sign_bit = if mul_bits == bits {
+            // If no extension took place, get the MSB.
```
Comment (LucasSte) [inline src/emit/mod.rs:5567]: You are not extracting the most significant bit. You are extracting the sign bit. If you have `int32 g = 4`, the most significant bit is the third, while the sign bit is the 32nd.

File: src/emit/mod.rs [hunk 15]
```diff
@@ -5505,6 +5505,200 @@ pub trait TargetRuntime<'a> {
...
+        let mul_by_zero = bin.builder.build_or(left_is_zero, right_is_zero, "");
+
+        let ok_operation =
+            bin.builder
+                .build_or(sim_operands_and_positive_res, diff_operands_and_neg_res, "");
+
+        // Here, we disregard the last rule mentioned above by oring with mul_by_zero
+        bin.builder.build_conditional_branch(
+            bin.builder.build_or(ok_operation, mul_by_zero, ""),
+            //  bin.context.bool_type().const_all_ones(),
```
Comment (LucasSte) [inline src/emit/mod.rs:5617]: Please, remove commented out code.

File: src/emit/mod.rs [hunk 16]
```diff
@@ -5505,6 +5505,200 @@ pub trait TargetRuntime<'a> {
...
+            "",
+        );
+
+        let res = bin.builder.build_load(o, "mul");
+
+        bin.builder
+            .build_int_truncate(res.into_int_value(), mul_type, "")
+    }
+
+    // Utility function to extract the MSB of an IntValue
```
Comment (LucasSte) [inline src/emit/mod.rs:5687]: There is a confusion here. The MSB is not the same as the sign bit. Please, use `SignBit` instead of MSB.

File: tests/solana_tests/primitives.rs [hunk 17]
```diff
@@ -599,6 +604,215 @@ fn truncate_uint(n: &mut U256, width: usize) {
...
+        .replace("intN", &format!("int{}", width));
+        let mut contract = build_solidity(&src);
+
+        contract.constructor("test", &[]);
+
+        // Generate the max value to fit in N bits
+        let limit = BigInt::from(2_u32).pow(width - 1).add(1_u32);
```
Comment (LucasSte) [inline tests/solana_tests/primitives.rs:734]: The bounds for positive and negative numbers are not equal. If your integer is 32-bit wide, for instance, the accepted range of integers is `[-2^31, 2^31-1]`. I would suggest you to add a test to test the boundaries of each case. This would be:

- Test that a result of `-2^31` does not overflow, but a result of `-2^31-1` do overflow.
- Test that a result of `2^31` does not overflow, but a result of `2^31` do overflow.

I used 32 bits as an example here. Please, use the bit widths that are convenient for your use-case.

File: tests/solana_tests/signature_verify.rs [hunk 18]
```diff
@@ -46,7 +46,7 @@ fn verify() {
 
     vm.constructor("foo", &[]);
 
-    let mut csprng = rand::thread_rng();
+    let mut csprng = rand_07::thread_rng();
```
Comment (LucasSte) [inline tests/solana_tests/signature_verify.rs:49]: What is the difference between version 0.7 and 0.8?

**Review by LucasSte — CHANGES_REQUESTED** [review]
Your code improved a lot from your first PR!

File: tests/solana_tests/primitives.rs (hunk 4 above)
Comment (salaheldinsoliman) [reply tests/solana_tests/primitives.rs]: Because that will waste time testing intrinsic LLVM functions, the work starts from 72 bits.
There are 4 tests like this so that will be a lot of unnecessary testing
